transports/janus_pfunix.c

```diff
@@ -0,0 +1,532 @@
+	}
+	janus_mutex_unlock(&clients_mutex);
+	/* Convert to string and enqueue */
+	char *payload = json_dumps(message, JSON_INDENT(3) | JSON_PRESERVE_ORDER);
+	json_decref(message);
+	g_async_queue_push(client->messages, payload);
+	/* Notify the thread there's data to send */
+	char c;
+	write(write_fd[1], &c, 1);
+	return 0;
+}
+
+void janus_pfunix_session_created(void *transport, guint64 session_id) {
+	/* We don't care */
+}
+
+void janus_pfunix_session_over(void *transport, guint64 session_id, gboolean timeout) {
+	/* We only care if it's a timeout: if so, close the connection */
+	if(transport == NULL || !timeout)
+		return;
+	/* FIXME Should we really close the connection in case of a timeout? */
+	janus_pfunix_client *client = (janus_pfunix_client *)transport;
+	janus_mutex_lock(&clients_mutex);
+	if(g_hash_table_lookup(clients, client) != NULL && client->fd > 0) {
+		/* Shutdown the client socket */
+		shutdown(client->fd, SHUT_WR);
+	}
+	janus_mutex_unlock(&clients_mutex);
+}
+
+
+/* Thread */
+void *janus_pfunix_thread(void *data) {
+	JANUS_LOG(LOG_INFO, "Unix Sockets thread started\n");
+
+	int fds = 0;
+	struct pollfd poll_fds[1024];	/* FIXME Should we allow for more clients? */
+	char buffer[4096];
+
+	while(g_atomic_int_get(&initialized) && !g_atomic_int_get(&stopping)) {
+		/* Prepare poll list of file descriptors */
+		fds = 0;
+		/* Writeable monitor */
+		poll_fds[fds].fd = write_fd[0];
+		poll_fds[fds].events = POLLIN;
+		fds++;
+		if(pfd > 0) {
+			/* Janus API */
+			poll_fds[fds].fd = pfd;
+			poll_fds[fds].events = POLLIN;
+			fds++;
+		}
+		if(admin_pfd > 0) {
+			/* Admin API */
+			poll_fds[fds].fd = admin_pfd;
+			poll_fds[fds].events = POLLIN;
+			fds++;
+		}
+		/* Iterate on available clients, to see if we need to POLLIN or POLLOUT too */
+		janus_mutex_lock(&clients_mutex);
+		GHashTableIter iter;
+		gpointer value;
+		g_hash_table_iter_init(&iter, clients_by_fd);
+		while(g_hash_table_iter_next(&iter, NULL, &value)) {
+			janus_pfunix_client *client = value;
+			if(client->fd > 0) {
+				poll_fds[fds].fd = client->fd;
+				poll_fds[fds].events = POLLIN;
+				if(g_async_queue_length(client->messages) > 0) {
+					poll_fds[fds].events = POLLOUT;
+				}
+				fds++;
+			}
+		}
+		janus_mutex_unlock(&clients_mutex);
+
+		/* Start polling */
+		int res = poll(poll_fds, fds, 10000);
+		if(res == 0)
+			continue;
+		if(res < 0) {
+			JANUS_LOG(LOG_ERR, "poll() failed\n");
+			break;
+		}
+		for(int i=0; i<fds; i++) {
+			if((poll_fds[i].revents & POLLOUT) == POLLOUT) {
+				/* Find the client from its file descriptor */
+				janus_mutex_lock(&clients_mutex);
+				janus_pfunix_client *client = g_hash_table_lookup(clients_by_fd, GINT_TO_POINTER(poll_fds[i].fd));
+				if(client != NULL && client->fd > 0) {
+					char *payload = NULL;
+					while((payload = g_async_queue_try_pop(client->messages)) != NULL) {
+						int res = write(client->fd, payload, strlen(payload));
+						/* FIXME Should we check if sent everything? */
```

Comment:
not if we use SOCK_DGRAM, because it's all or nothing, so bonus nachos. But you also want to do the do-while-eintr dance here.